File: Top_Interview_150/Matrix/python/54.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        n = len(matrix[0])

        # if m == 1 or n == 1:
        #     return matrix

        count = 1
        i = 0 if m > 1 else 1
        j = 1 if m > 1 else 0
        height_step = line_step = 1
        result = matrix[0][0]

        flag = m > 1

        while count < m * n:
            if flag:
                j += line_step
                if j == 0 or j == n - 1:
                    flag = False
                    line_step = - line_step
            else:
                i += height_flag

File: Top_Interview_150/Matrix/cpp/54.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix)
    {
        int m = matrix.size();
        int n = matrix[0].size();

        vector<int> result;
        int left = 0, up = 1, right = n, down = m;
        int i = 0, j = -1;
        while (result.size() < m * n)
        {
            j++;
            while (result.size() < m * n && j < right)
                result.push_back(matrix[i][j++]);
            j--;
            right--;

            i++;
            while (result.size() < m * n && i < down)
                result.push_back(matrix[i++][j]);
            i--;
            down--;

            j--;
            while (result.size() < m * n && j >= left)
                result.push_back(matrix[i][j--]);
            j++;
            left++;

            i--;
            while (result.size() < m * n && i >= up)
                result.push_back(matrix[i--][j]);
            i++;
            up++;
        }

        return result;
    }
};